Update parented_ptr

timrae opened the pull request:
Following on from the discussion in #1142, having the copy constructor available in the `parented_ptr` class gives an incorrect feeling that it could be used in the same way as a `shared_ptr`, which would lead to dangling pointers if the pointer can outlive its parent.

To address this problem we introduce a second class `parented_weak_ptr` that wraps the raw pointer in a QPointer. Since @rryan pointed out that QPointer has some overhead, we keep the old `parented_ptr` which doesn't use QPointer, but change it so that it can only be used with move semantics, making it safer than the version in #1142 in that the compiler will detect some (but not all) possible abuses.

Basically the intention with this class is that the "owner" (of the `parented_ptr`) and the "parent" are the same object. In such cases the overhead of `QPointer` is unwarranted:

```c++
class ParentClass {
  private:
    parented_ptr<T> m_pChild = std::make_parented<T>(this);
};
```

The changes in this PR are working well in practice, as you can see in c9eaaf47d142d90cb538baedb27750e6eb31d71e.

Note that I [initially](https://gist.github.com/timrae/f6cca4af6a9ad8d9d45e76b0e23ab909) tried to make a version of `parented_ptr` that inherits from `unique_ptr` to minimize duplicated code, but I couldn't get that to work, as it was complaining about the deleted copy constructor when initializing `m_pLibraryControl` in c9eaaf47d142d90cb538baedb27750e6eb31d71e. If someone can figure out the trick how to get that to work, I think it would be worth changing to such an implementation.

daschuer commented:
A parented_unique_ptr pointer holds the error already in its name. The pointer cannot be unique, because it is actually copy of the pointer stored in the widget tree. So it does not make sense to have no copy feature. 

What could be useful is a light weight "parented_ptr", that documents only 
"This pointer is owned by the parent object" 
and "parented_weak_ptr", that tells us 
"This pointer is owned by the parent object, and becomes null if the object is deleted. 

> Following on from the discussion in #1142, having the copy constructor available in the parented_ptr class gives an incorrect feeling that it could be used in the same way as a shared_ptr, which would lead to dangling pointers if the pointer can outlive its parent.

I think with a parented_weak_ptr this issue can be addressed and the rule: 
* Use parented_weak_ptr if a pointer to an object is stored in an object that can outlive the owning object tree. 

What do you think? 

By the way: 
In QT5 the QPointer is essentially a QWeakPointer: 
https://github.com/qt/qtbase/commit/6f0f9f69288925ef423c542ef5eb7302a5431867#diff-01ee4fa3469b2f5d04a50bdcb97ed6b4
which is much more lightweight now. 
But there is still bit overhead compared to a plain pointer, because the reference counting using is initialized in an extra heap memory.

timrae commented:
> A parented_unique_ptr pointer holds the error already in its name. The pointer cannot be unique, because it is actually copy of the pointer stored in the widget tree.

Well I'm not strongly attached to the `parented_unique_ptr` name, but I don't think your argument is completely correct. What matters is that the *data structure* is unique, not necessarily the *underlying pointer*. According to your reasoning, `std::unique_ptr` is also mistaken because it has a `get()` and a `reset()` function, and thus makes no guarantees about the uniqueness of the underlying pointer.

> It does not make sense to have no copy feature.

I disagree. Making a copy of a smart pointer is only supposed be done when ownership is to be shared. In the situations where this data structure is intended to be used, I think it will basically always be a mistake to make a copy, so disabling the copy constructor increases the safety.

> What could be useful is a light weight "parented_ptr", that documents only
"This pointer is owned by the parent object"
and "parented_weak_ptr", that tells us
"This pointer is owned by the parent object, and becomes null if the object is deleted.

Sure, I don't mind changing the names so that in this PR: `parented_ptr` -> `parented_weak_ptr` and `parented_unique_ptr` -> `parented_ptr`. However, I think we should continue to block copying of the `parented_ptr` for the reason I mentioned above.

> (and) the rule:
* Use parented_weak_ptr if a pointer to an object is stored in an object that can outlive the owning object tree.

Yes, I would take the rule one step further though, and say:

* Use `parented_weak_ptr` unless it is obvious that object can never outlive the owning object tree.

timrae commented:
> By the way: In QT5 the QPointer is essentially a QWeakPointer

In that documentation it says that `QPointer` is *obsolete* now. In that case, should we switch from `QPointer` to `QWeakPointer` here as well?

daschuer commented:
>>    By the way: In QT5 the QPointer is essentially a QWeakPointer

> In that documentation it says that QPointer is obsolete now. In that case, should we switch from QPointer to QWeakPointer here as well?

They have reverted the decision later and removed the obsolete marks. 
https://github.com/qt/qtbase/commit/b8773165d76e0d5d46287d92f9d6bdbbd2110180

And her is why: 
http://www.macieira.org/blog/2012/07/continue-using-qpointer/

daschuer commented:
>>  A parented_unique_ptr pointer holds the error already in its name. The pointer cannot be unique, because it is actually copy of the pointer stored in the widget tree.

> Well I'm not strongly attached to the parented_unique_ptr name, but I don't think your argument is completely correct. What matters is that the data structure is unique, not necessarily the underlying pointer. According to your reasoning, std::unique_ptr is also mistaken because it has a get() and a reset() function, and thus makes no guarantees about the uniqueness of the underlying pointer.

Yes that makes sense. 

Maybe the reason why I feel uncomfortable is that a std::unique_ptr represents the unique ownership itself which is moved (only) along with this pointer. That is the reason because std::unique_ptr has no copy constructor. This avoids effectively to have two std::unique_ptr for the same object. 

This is not the case here and adopting the this concept could be misleading. Our parented_ptr is basically a non owning pointer, like a raw or a weak pointer that can copied freely without violating any smart pointer rules, because it does not transfer the ownership.  

> * Use parented_weak_ptr unless it is obvious that the object can never outlive the owning object tree.

OK 

If we remove the copy constructor from the parented_ptr, we cannot pass it as a function argument anymore. I this desired?

timrae commented:
OK I've force pushed a new commit with the names that you suggested.

> Our parented_ptr is basically a non owning pointer, like a raw or a weak pointer that can copied freely without violating any smart pointer rules, because it does not transfer the ownership.

I actually strongly disagree with this... Let's say we have a parented_ptr `p_parented`. The idea with this class is that the owner of `p_parented` should be the same object as `p_parented->parent()`. Ownership and lifetime here are both very clear and 100% related. See the example code in the PR description if it's not 100% clear what I mean.

As soon as a copy of `p_parented` (let's call it `p_parented_copy`) is made, the above assumption is violated, and in most cases it becomes difficult to reason about whether `p_parented_copy` is valid or not, as `p_parented_copy->parent()` is not the same object as the owner of `p_parented_copy`.

So just like a `unique_ptr`, `parented_ptr` should never be copied. The only way it's different from a `unique_ptr` is that ownership is not automatically transferred when you `std::move()` it, as you need to additionally reassign the parent after moving.

timrae commented:
> If we remove the copy constructor from the parented_ptr, we cannot pass it as a function argument anymore. Is this desired?

Yes, it should not never be copied into another function. If the function is not to take ownership then we pass in a raw pointer via `p_parented.get()`. If the function **is** to take ownership then we pass it in with `std::move()`. Additionally, if we `std::move()` it then we should reassign the parent to the new owner. I can't imagine this situation coming up very often though, it would be more common to create the object with no parent via `std::make_unique()`, then `std::move()` it into a function, and create the `parented_ptr` from that, as we discussed in #1142

timrae commented:
I edited the PR title and description to reflect the new naming

rryan commented:
> @timrae: If the function is not to take ownership then we pass it a raw pointer via p_parented.get(). 

:+1: to passing raw pointers around being the default for code that does not take ownership or does not need to store a weak reference. We will frequently have to do that anyway, since one awkward detail with smart pointers is that you if you write all your code that deals with a class to take smart pointers, then the class itself cannot pass itself to functions that take a smart pointer because it only has a raw pointer to itself (```this```). 

> @timrae: If the function is to take ownership then we pass it in via std::move(p_parented). Additionally, if we do std::move() it then we should reassign the parent to the new owner (i.e. p_parented->setParent(this)).

We could make something like this: 
```c++
template <typename T>
typename remove_reference<T>::type&& move_parent(QObject* parent, T&& child) {
  child->setParent(parent):
  DEBUG_ASSERT(child->parent() == parent);
  return static_cast<typename remove_reference<T>::type&&>(arg);
}
```
It's annoying that setParent can fail though.

timrae commented:
Hmm do you think that function would actually need to be used in practice though? I can't really think of any examples where you'd need to set a parent then change it... What are the possible failure reasons for setParent ()?

rryan commented:
> Hmm do you think that function would actually need to be used in practice though? I can't really think of any examples where you'd need to set a parent then change it...

No -- I don't think we do it in Mixxx so we probably don't need it. All the examples of setParent in the codebase don't seem to do it. They're all setting a parent where the object previously has none. 

> What are the possible failure reasons for setParent()?

I think setParent only fails if the parent and new child's ```QObject::thread()``` do not match. 

The only place I can think of where we're regularly changing parents is for Singletons in the skin system. And since those are QWidgets, they are required to live on the Qt main thread so the parent and child will always have matching thread properties. So I think setParent can't fail in that case.

timrae commented:
Ok maybe we cross that bridge when we come to it then. If we do need reparenting in the API we could initially limit it to QWidgets, or in the future when the language eventually supports concepts maybe we could have a reparentable concept.

rryan commented:
> Ok maybe we cross that bridge when we come to it then.

SGTM

timrae commented:
By the way, I'm in bed at the moment so can't check, but if we have a parented_ptr `p` which has a child parented_ptr `c` as a member variable and we do std:: move on `p`, does `c->parent()` still work as expected?

daschuer commented:
> I actually strongly disagree with this... Let's say we have a parented_ptr p_parented. The idea with this class is that the owner of p_parented should be the same object as p_parented->parent().

This view changes what parented_ptr is by 180°. 

Before it was just decoration for a raw pointer that says "This is a non owning pointer, to an object that is owned by a parent object. Now you are about to change it to: "This is a pointer to an object that is owned by >this< as a parent. 

I am afraid the new view is quite unhandy in many MVC situations, where a controller object constructs the widget tree. In this situation the parent object is not equal to this. For my feeling this is far more common than an inherited widget that constructs its own parents. 

You may look for example at the latest changes Joan has introduced here:
https://github.com/daschuer/mixxx/pull/15/files#diff-ab47b00ac7c66c3ae2f629389146f163R688

> If the function is not to take ownership then we pass it a raw pointer via p_parented.get(). If the function is to take ownership then we pass it in via std::move(p_parented). Additionally, if we do std::move() it then we should reassign the parent to the new owner (i.e. p_parented->setParent(this)).

The issue with this view is that a parented_ptr() does currently not own the object. So it cannot transfer the ownership. I am afraid this will finally more confusing than help, because it looks like an owning pointer, which is wrong or artificially enforced. 

IMHO the current "parented_ptr" works already pretty well. It asserts an object has a parent, it allows us to hide plain new() calls and store plain pointers as member variables. 
The parented_weak_ptr adds the live time tracking feature, which is also nice, and fixed the initial issue of this PR. 
 
I am not sure what the value of a "parented_this_ptr" is in the context of readability.

timrae commented:
> This view changes what parented_ptr is by 180°.

Yes, that's why I initially introduced it as a new class.

> I am afraid the new view is quite unhandy in many MVC situations, where a controller object constructs the widget tree. In this situation the parent object is not equal to this.

Why not use in parented_weak_ptr in such cases? We're about to move to qt5 right? I'm already using qt5 on OSX because 4 is no longer supported by homebrew.

> IMHO the current "parented_ptr" works already pretty well.

"Pretty well" is not good enough. When I read the code I want to immediately see that there are no memory leaks or dangling pointers without thinking about it too much. This PR gives us that, parented_ptr does not.

timrae commented:
Maybe `parented_ptr_unsafe` would be a better name for it, kind of like in the rust unsafe mode.

daschuer commented:
> Why not use in parented_weak_ptr in such cases? We're about to move to qt5 right? I'm already using qt5 on OSX because 4 is no longer supported by homebrew.

An predicting ownership pointer might be confusing for people who are knowing QT very well. 
It looks like it has the ownership but actually we only try to beat the Qt API where we can't.

We can consider to use always parented_weak_ptr 2.1 is released. But I am not feeling well to introduce a useless heap allocation and reference counting to clean it up, just because of an self made programming pattern.

> "Pretty well" is not good enough. When I read the code I want to immediately see that there are no memory leaks or dangling pointers without thinking about it too much. This PR gives us that, parented_ptr does not.

A parented_weak_ptr allown gives us that. 

A non copy-able parented_ptr is IMHO a source of confusion. I am a bit afraid that the code predict more than it actually is. We have to consider always refactoring, which may brake the nice rules, If they are not enforced by a compiler.

daschuer commented:
> Maybe parented_ptr_unsafe would be a better name for it, kind of like in the rust unsafe mode.

Or just parented_raw_ptr?

timrae commented:
> We can consider to use always parented_weak_ptr after 2.1 is released. But I am not feeling well to introduce a useless heap allocation and reference counting to clean it up, just because of an self made programming pattern.

> A parented_weak_ptr alone gives us that.

You seem to be contradicting yourself in the same post... I agree with the first point, which makes the second point redundant. I think we agree that `parented_weak_ptr` is good, but not enough. So now we need to make the new unsafe/raw pointer as safe and useful as we can, and we do this by introducing clear and easily enforceable rules on its usage. Preferably the rules should be enforced by the compiler when this is possible.

The rule you previously suggested was:

* Only use `parented_raw_ptr` when the parent outlives the pointer.

While this is correct and sufficient to prevent dangling pointers, it's not easily enforceable. You may have to hunt through the code to track the lifetime of the pointer and the parent, and then come to a conclusion about whether or not it's safe. If you allow the pointer to be copyable then you drastically increase this work load, drastically increase the chance of making a mistake, and drastically increase the chance that the code will be broken later by someone else who didn't understand it as well as you.

*If your use case is sufficiently complex that you need to copy the pointer, then you should be using `parented_weak_ptr`*

timrae commented:
> I am afraid the new view is quite unhandy in many MVC situations, where a controller object constructs the widget tree. In this situation the parent object is not equal to this. For my feeling this is far more common than an inherited widget that constructs its own parents.

> You may look for example at the latest changes Joan has introduced here:
https://github.com/daschuer/mixxx/pull/15/files#diff-ab47b00ac7c66c3ae2f629389146f163R688


If the line below is really required then it should be using `parented_weak_ptr`:

> `m_panes[paneId] = pStack.get(); // BAD! Taking ownership of a raw pointer`

If not, we can easily rewrite the code to use `parented_raw_ptr`:

```c++
parented_raw_ptr<QWidget> BasePlaylistFeature::createPaneWidget(KeyboardEventFilter* pKeyboard, int paneId, QWidget* parent) {
    auto pStack = make_parented_raw<WLibraryStack>(parent);
    //m_panes[paneId] = pStack.get();
    
    auto edit = make_parented_raw<WLibraryTextBrowser>(pStack.get());
    edit->setHtml(getRootViewHtml());
    edit->setOpenLinks(false);
    edit->installEventFilter(pKeyboard);
    connect(edit.get(), SIGNAL(anchorClicked(const QUrl)),
            this, SLOT(htmlLinkClicked(const QUrl)));
    m_browseIndexByPaneId[paneId] = pStack->addWidget(edit.get());
    
    auto pTable = LibraryFeature::createPaneWidget(pKeyboard, paneId, pStack.get());
    m_tableIndexByPaneId[paneId] = pStack->addWidget(pTable.get());

    return std::move(pStack);
}
```

daschuer commented:
Ok, I think we have a conclusion that it is a good idea to introduce the parented_raw_ptr. Fine.

> The rule you previously suggested was:
> *    Only use parented_raw_ptr when the parent outlives the pointer.

Not exact. How about this: 

* Use parented_weak_ptr if the pointer can outlive the owning parent object, or if the relationship between the object holding the pointer and the object owning the object is not obvious.

--- 

In @jmigual code, luckily panes is already ```QHash<int, QPointer<WLibraryStack> > m_panes;``` this could become ```QHash<int, parented_weak_ptr<WLibraryStack> > m_panes;``` after this is merged. 

The issue here is that this is not obvious when reading ```m_panes[paneId] = pStack.get();```
so we should introduce here something like ```m_panes[paneId] = pStack.toParentedWeak();``` and prevent the old call with some c++ magic. 

---

```return std::move(pStack);``` the move is redundant here because the 

---

QPointer qt5 has a swap() function, which allows us to move a  ```parented_weak_ptr``` 
this could be handy to avoid reference counting, when passing the pointer. 

---

I am starting to understand why you do not want to have the parented_raw_ptr copy-able. 

This will encourage us to use a parented_weak_ptr once it is handed out.  
This will work if we switch back to to a mode that we use plain pointer for function parameters like you did in your example. This follows the rule that is true for all std:: pointers "use get() to borrow the pointer"

We only have the issue, that a new QObject stores the parent pointer. This is a kind of weak pointer, because the parent outlived the child and so this pointer is always valid, or null. Because of this implied safety, it would be OK for me to use a plain pointer here. 

This will revert our decision here: https://github.com/daschuer/mixxx/pull/15#discussion_r97743688
but I agree that the overall safety and readability is improved.

timrae commented:
> Ok, I think we have a conclusion that it is a good idea to introduce the parented_raw_ptr. Fine.

Great, I just pushed the name change.

> Not exact. How about this:

> * Use parented_weak_ptr if the pointer can outlive the owning parent object, or if the relationship between the object holding the pointer and the object owning the object is not obvious.

Yes, sounds good to me...

daschuer commented on src/util/parented_raw_ptr.h:
```diff
@@ -0,0 +1,90 @@
+
+    T* get() const {
+        return m_pObject;
+    }
```
please also add the ```toParentedWeak()``` function.

timrae commented:
> please also add the toParentedWeak() function.

I don't really see an advantage to using this function though... If the pointer is to be shared I feel you should just use `make_parented_weak<T>(p)`

daschuer commented:
> If the pointer is to be shared I feel you should just use make_parented_weak<T>(p) 

this will be less efficient, especially if we are still in QT 4 where a QPointer has no swap()

It is basically all about our example code above. I think after merging this, it should look like: 

```c++
parented_ptr<QWidget> BasePlaylistFeature::createPaneWidget(
            KeyboardEventFilter* pKeyboard, int paneId, QWidget* parent) {
    auto pStack = make_parented<WLibraryStack>(parent);
    m_panes[paneId] = pStack.toParentedWeak();
    
    auto edit = make_parented<WLibraryTextBrowser>(pStack.get());
    edit->setHtml(getRootViewHtml());
    edit->setOpenLinks(false);
    edit->installEventFilter(pKeyboard);
    connect(edit.get(), SIGNAL(anchorClicked(const QUrl)),
            this, SLOT(htmlLinkClicked(const QUrl)));
    m_browseIndexByPaneId[paneId] = pStack->addWidget(edit.get());
    
    auto pTable = LibraryFeature::createPaneWidget(pKeyboard, paneId, pStack);
    m_tableIndexByPaneId[paneId] = pStack->addWidget(pTable.get());
    
    return pStack;
}
```

Could you also add a move constructor to parented_weak_ptr using QPointer::swap() for the QT 5 case?

timrae commented:
When are we switching to QT5?

rryan commented:
> When are we switching to QT5?

Mixxx 2.2

timrae commented:
So the dev branch can be switched over as soon as the release branch is
forked? Aka last week ;)

On 1 Feb. 2017 16:17, "RJ Skerry-Ryan" <notifications@github.com> wrote:

> When are we switching to QT5?
>
> Mixxx 2.2
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mixxxdj/mixxx/pull/1161#issuecomment-276588954>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACsA4s06Idez2Fu-Dzs6lHMqJ79smqQjks5rYDFxgaJpZM4LwzZN>
> .
>

daschuer commented:
Just sounding the move constructor using QT5 functions by 
```c++
#if QT_VERSION >= QT_VERSION_CHECK(5, 0, 0)

#endif
```
And we can continue here independently.

timrae commented:
I don't understand how this:
`m_panes[paneId] = pStack.toParentedWeak();`

Is supposed to be different or better than this:

`m_panes[paneId] = parented_weak_ptr<WLibraryStack>(pStack.get());`

timrae commented:
Is it just to be less verbose? I don't like the `to` in the function name... `getWeakReference()` sounds more descriptive to me

daschuer commented:
```to..``` is part of the Qt API whenever a conversion is done. ```get``` can be interpreted as a simple getter for a member. 
http://doc.qt.io/qt-5/qsharedpointer.html#toWeakRef

`m_panes[paneId] = parented_weak_ptr<WLibraryStack>(pStack.get());`
Will work the same. It just requires more brain and typing.

I have no strong opinion in this. 

Is there a way to fail 
`m_panes[paneId] = pStack.get();` 
which is a source of confusion? 
That is the main issue for me.

timrae commented:
Maybe we can omit the parented in the function name?

```c++
parented_weak_ptr<T> toWeakRef();
```

daschuer commented:
OK.

timrae commented:
I force pushed a new commit with the changes you requested, let me know if there were any misunderstandings...

Be-ing commented:
Would this be useful to other Qt applications? You may want to consider publishing it as a separate library.

timrae commented:
Good idea,  let's see how it works for us first though

daschuer commented on src/util/parented_raw_ptr.h:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    // Delete copy constructor and copy assignment operator
+    parented_raw_ptr(const T&) = delete;
```
Typo: 
parented_raw_ptr(const parented_raw_ptr<T>& p) = delete;

timrae replied:
Fixed, thanks

timrae commented:
@daschuer 
The 2.1 branch has been created now... can we switch to QT5 in master and remove the `#if QT_VERSION >= QT_VERSION_CHECK(5, 0, 0)` directives?

timrae committed:
Disable copying in "parented_ptr"
A new function toWeakRef() was introduced which can be copied freely in cases where the parented_ptr copy operations would ordinarily be used.

timrae commented:
@daschuer 
Idea:

How about changing `toWeakRef()` to just return an ordinary `QPointer<T>`, and getting rid of `parented_weak_ptr` entirely? The goals of these classes are to have no naked new statements or dangling references, and I feel we can achieve these goals entirely with `parented_raw_ptr<T>` and `toWeakRef()`.

The usage rules are then simple:

* If an object is to be owned by the QT object tree, create it with `make_parented_raw`.
* **Never** store the `parented_raw_ptr` in an object that outlives the parent of the object. If you need to do this then store the output of `toWeakRef()`
* If a function input/output should simply "borrow" a reference without storing it then use a raw pointer `T*` via `get()`
* If a function input/output needs to keep a reference to the object then use a weak reference via `toWeakRef()`.

daschuer commented:
> The 2.1 branch has been created now... can we switch to QT5 in master and remove the #if QT_VERSION >= QT_VERSION_CHECK(5, 0, 0) directives?

We still keep both branches in sync. Even if we switch to QT5, we should not brake QT4 builds until the QT5 version is settled, or any other strong reason.

daschuer commented:
> Idea: .. 

That would make the code more readable for a new contributor knowing Qt but not Mixxx. 

On the other hand, we have no guaranty, that the QPointer stored outside the object tree has a parent. 

I am unsure, if this is a critical question? Probably not, because the storing object of the QPointer can silently assume that the lifetime is tracked elsewhere. The same is true for a normal weak pointer. 
The difference here is that you cannot create a weak pointer without am owning shared pointer. 

T oughts?

timrae commented:
Yeah I think it's a non issue. During code review we check for memory leaks at the time of object creation, so having a QPointer member variable should not cause any red flags. Code review will be simple:

* If you have a naked new you might have a memory leak
* If you have a naked delete you might have a segfault
* If you have a raw pointer member variable you might have a dangling pointer
* If you dereference a QPointer without checking if it's null you might have a segfault

timrae commented:
@daschuer 
I've pushed the changes, I think this way is much more clean and intuitive. Could you please re-review? I think it's ready to merge now.

daschuer commented:
Typo: m_pObject == other.m_pObject
```
    bool operator== (const parented_ptr& other) const {
        return m_pObject == other.m_ptr;
    }
```
Why is it not part of the diff?

timrae commented:
> Why is it not part of the diff?

https://github.com/mixxxdj/mixxx/blob/master/src/util/parented_ptr.h

Because it's in master as well :-1:

timrae committed:
Fix typo in parented_ptr

daschuer commented:
LGTM! Thank you.

Be-ing commented:
I appreciate that you're taking the time to design a well thought out solution and not making quick hacks. :)

timrae commented:
Thank you 😁

timrae commented:
None of this would have happened without @daschuer's continual input

daschuer commented:
Discussing about all this, I have learned a lot, thank you. 
Now we should add the new rules to our guidelines. I will collect a draft into the following comment.

daschuer commented:
A first a bit awkward draft:

**Pointer Usage**

* When ever possible, avoid naked `new` and `delete` statements, use std::make_unique<T> std::make_shared<T> or make_parented_raw<T> to clarify the ownership just from the start. 
* Use smart pointers or parented QObjects to track the live time and ownership of an object.  
* Plain pointers are used for function arguments as non owning pointers that are not stored after returning from the function. An exception is the parent pointer passed for creating a QObject and stored as a non owning reference. 
* As per the previous rule, in general we can say that a class should not have raw pointers as member variables. Always wrap pointer member variables into an appropriated smart pointer. 
* If a pointer is to be managed by the QT object tree, and has a known parent at the time of creation, use make_parented_raw<T>
* If a pointer is not to be managed by the QT object tree, use std::make_unique<T> to create the object, or std::make_shared<T> if shared ownership is really necessary.
* If a pointer that will eventually be managed by the object tree needs to be created without a parent, create the object p with std::make_unique<T>, then later create a new variable using parented_ptr(p.release()) once the underlying object gets a parent.
* Never store the parented_raw_ptr in an object that outlives the parent of the object. If the lifetime is not clear then store the output of toWeakRef()

* Any exceptions to these guidelines must be clearly documented inline in the code

timrae commented:
I think the rules are too many and contain some redundant information. How about the following?

* Follow the isocpp [no naked new rule](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#es60-avoid-new-and-delete-outside-resource-management-functions). Instead of using `new` and `delete` statements, use one of the following "smart pointers" to convey ownership right from the start:

  - `std::unique_ptr` for exclusive ownership: e.g. `auto p_t = std::make_unique<T>(...)`
  - `std::shared_ptr` for shared ownership: e.g. `auto p_t = std::make_shared<T>(...)`
  - `parented_ptr` for ownership by the QT object tree: e.g. `auto p_t = make_parented<T>(...)`

 * A raw pointer `T*` can be freely passed into or returned from a function, but a function which receives a raw pointer should not hold onto the pointer outside the scope of the function (e.g. by storing it in a member variable). An exception is the parent pointer passed into the constructor of a `QObject` and internally stored as a non owning reference.

* `parented_ptr` requires the parent to be passed into the constructor at the time of creation. If a pointer that will eventually be managed by the QT object tree needs to be temporarily created without a parent, first create the object `p` with `std::make_unique`, and then later create a new variable using `parented_ptr(p.release())` once the underlying object gets a parent.

* Never store a `parented_ptr` in an object that outlives the parent. If the lifetimes are not clear then convert the `parented_ptr` to a weak reference (via `toWeakRef()`) and store that.

* Any exceptions to these guidelines must be clearly documented inline in the code

daschuer commented:
Yes much better. :-) I will put it to our guidelines.

daschuer commented:
https://www.mixxx.org/wiki/doku.php/coding_guidelines#pointer_object_live_time_ownership

timrae commented:
I was reading through the Cpp Core Guidelines again today (the section on [smart pointers](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#rsmart-smart-pointers) to be specific), to see if there's anything mentioned there that we forgot about... 

It seems that our wording is a bit too soft on passing raw pointers / references (i.e. you "can" do it), whereas it should be more strong (you "must" do it if there is no ownership semantics). Specifically, passing a smart pointer by reference is usually a big no-no, as such semantics is reserved for reassigning the smart pointer to a different object:

 https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#r33-take-a-unique_ptrwidget-parameter-to-express-that-a-function-reseats-thewidget

I think we can reword that text a bit more, and maybe drop another link to the guidelines.

One interesting point of difference between the Cpp Core Guidelines and the Google Style Guide is that Google says: "Always use a pointer over a non-const ref as input to a function", and CCG says: "Always use a reference over a pointer when the object is known to be non-null".

The former is just for readability reasons, whereas the latter is much more syntactically clear and can prevent unnecessary null checking as well as accidental dereferencing of null pointers. I.e. code that follows the CCG is generally going to be faster, safer, and clearer.

@rryan 
What do you think?

daschuer commented:
It is a myth, that a reference cannot be null. 

You get a straight segfault with the following code: 

```
void do_thomething(int& i) {
    i++;
}

void main() {
    int* pi = nullptr;
    do_thomething(*pi);
}
```

In Mixxx we follow the Google view and I like it very much because it distinguishes well local access from access that effects outside code. 

I will improve our rules  for "should be" and not passed by reference.

daschuer commented:
- Pass raw pointer T* if no ownership is transfered. A function which receives a raw pointer should not hold onto the pointer outside the scope of the function (e.g. by storing it in a member variable). An exception is the parent pointer passed into the constructor of a QObject and internally stored as a non owning reference. Do not pass smart pointers by reference.

timrae commented:
> You get a straight segfault with the following code

I think your understanding is incorrect. You get undefined behaviour as soon as you dereference the null pointer. In general it will segfault before it even enters the function. There is no *valid* C++ code that gives a null reference . They discuss it on that page somewhere.

Your argument about distinction is also not compelling. We are talking only about non const pointer vs non const reference. If the function doesn't modify the outside world then pass it const variables, that's what the keyword is for. References as a language feature are, by design, intended to provide clearer and easier syntax for non null pointers. Why not use them for that?

The Google rule is outdated and irrelevant in the modern world in my opinion. I'm willing to bet money they abandon it within the next 5 years.

daschuer commented:
The following code does not segfault: 

```
void do_thomething(int& i) {
    if (&i != nullptr) {
        i++;
    }
}

void main() {
    int* pi = nullptr;
    do_thomething(*pi);

    pi = new int();
    do_thomething(*pi);
}
```

> Your argument about distinction is also not compelling. We are talking only about non const pointer vs non const reference. If the function doesn't modify the outside world then pass it const variables, that's what the keyword is for. References as a language feature are, by design, intended to provide clearer and easier syntax for non null pointers. Why not use them for that?

In my real live, I code c# where everything is basically a pointer with reference syntax. It is a pain, that everything can be changed from everywhere. The const key word of c++ is a great improvement, but it does not help at the calling code. 

```
do_thomething(&i); // we passed a pointer, be carefull, the i will be most likely changed 
do_thomething(i); // when we allow non const references, we do not know, is i changed after the call 
```

Qt follows the same rule and all the Mixxx code is like that. So lets stick with it until the 5 Years have passed and we know better.

timrae commented:
> The following code does not segfault

Maybe not right now on your compiler with the default optimisation settings but it might. As soon as you dereference the null pointer you are not using valid C++ and you will get undefined behaviour. The compiler can do anything it likes including make demons come out of your nose.

rryan commented:
> Your argument about distinction is also not compelling. We are talking only about non const pointer vs non const reference. If the function doesn't modify the outside world then pass it const variables, that's what the keyword is for. References as a language feature are, by design, intended to provide clearer and easier syntax for non null pointers. Why not use them for that?

Why is it not compelling? There is no distinction, syntactically, between a local variable and a reference. So if you are looking at a piece of code you can't tell whether mutations "escape" the scope they are in without keeping track of whether the object is a reference or not in your head. This hurts readability. The pointer dereference operator gives you a good hint that your mutations last beyond the function call.

IMO non-const references are a dangerous mis-feature of C++. I'd strongly prefer we keep our rule rather than start using non-const references. I wish there was a better syntax for passing a non-null pointer than references, but there isn't.

timrae commented:
> do_thomething(&i); // we passed a pointer, be carefull

For me this just demonstrates my point... What I read here is that do_thomething(&i) may set i to nullptr. In 99% of code this isn't what you want. Why would you want to make your code less safe just to avoid looking at the function declaration?

daschuer commented:
do_thomething(&i) cannot set i to nullptr. It can only change the value of i. 

> I wish there was a better syntax for passing a non-null pointer than references, but there isn't.

The CppCoreGuide has: 
https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Ri-nullptr
https://github.com/Microsoft/GSL/blob/4e8f95b418ae96c802de57a2c041f16725a3d8ba/include/gsl/gsl#L72

timrae commented:
> do_thomething(&i) cannot set i to nullptr.

Ooops yes in that case you're right because the pointer is an rvalue. Although in the general case when you pass in an lvalue my argument would apply.

> if you are looking at a piece of code you can't tell whether mutations "escape" the scope they are in without keeping track of whether the object is a reference or not in your head. This hurts readability.

We are talking primarily about function arguments here... If your function is so long that you can't see the whole thing on your monitor, or you have so many arguments that you can't remember which is supposed to be modified then you've already failed on the readability front IMO :-p 

> I wish there was a better syntax for passing a non-null pointer than references, but there isn't.

At least on this we agree. 

> The CppCoreGuide has (`not_null<T>`)

Yes I was planning on suggesting this as a middle ground if we were not able to agree on the more compact `T&`. @rryan what do you think?

rryan commented:
> We are talking primarily about function arguments here... If your function is so long that you can't see the whole thing on your monitor, or you have so many arguments that you can't remember which is supposed to be modified then you've already failed on the readability front IMO :-p

Yea, I agree.

> Yes I was planning on suggesting this as a middle ground if we were not able to agree on the more compact T&. @rryan what do you think?

I guess it works, but It's kind of onerous for what I see as a non-issue in practice.

timrae commented:
If dereferencing a nullptr is a non-issue then I don't know what is... How is one supposed to know when reading the code where the null check is supposed to go? I see a lot of code with either too many or too few null checks. The former is wasteful and the latter is just plain dangerous.

timrae commented:
I agree that it is tedious to put `not_null<T>*` in all the function declarations. But my perspective is that this should be solved using references, and the extra verbosity is a result of your preference to increase readability in other areas (as discussed above), which I personally see as much more of a non-issue than potential undefined behavior or slower code.

rryan commented:
All pointer function arguments should be null checked before dereference, of course. What I mean is I don't think it's very common that we write a function that takes pointer arguments where we know the argument's going to be non-null.

timrae commented:
> All pointer function arguments should be null checked before dereference, of course. What I mean is I don't think it's very common that we write a function that takes pointer arguments where we know the argument's going to be non-null.

Right, yeah I'm talking about cases like the below example, where `baz()` would be required to do a redundant null check before it uses the input argument.

```c++
foo(T* p_t) {
    if (p_t) {
        p_t->bar();
        baz(p_t);
    }
}
```

I agree that it's not a hugely common occurrence though, and the below is far more comon:

```c++
foo(T* p_t) {
    baz(p_t);
}
```

in this case we can delay the check until we're inside `baz()`. If we're happy to deal with the tiny performance penalty in the first case then we can just have the rule that you must do a null check the first time you dereference a raw pointer or `QPointer` in a function.

daschuer commented:
@timrae: I think your interpretation of the c++ standard how to use `T&` is correct. It is intended to use the code like that.

The only problem with it is that it solves issues by introducing new issues. Both views are valid, and it is IMHO a matter of taste what to use, if we decide for one. A mixture of both would make the issue worse. 

Since our current code and the surrounding code does not use T&, we should stick with it. 

--- 

Some details. 

No one can enforce that a `T&` references a valid object. The underlying assembler implementation is petty much the same as using a pointer. It is just a rule that can be broken in the same way we can forget to check a pointer against null.

The rule that applies for `T&` and `const T&` is " do not create it by dereference a pointer that can be null.  

```c++
    // wrong!
    do_thomething(*pi);
    
    // right  
    if (pi != nullptr) {
         do_thomething(*pi);
    }
```

On the other hand, introducing `T&`` brakes the assumption that an access to an object without a dereferencing it effects only the local scope. 

```c++
    void do_thomething_more(T a, T& b, const T& c, T&& d, T* e) {
        ++a;  // effects the local copy     
        ++b;  // effects a in the calling code !  
        ++c;  // will not compile  
        ++d;  // effects moved object, d is invalid in the calling code 
        ++(*e); // effects a in the calling code !  
    ...
    }
```

b is the source of confusion, because it cannot be distinguish from the other cases, where the call only effects the local objects.

You argue: Look to the function arguments, but consider that we do the majority of code reviews on unified diffs, where the function header is out of sight. 
It is hard enough, to figure out what the code does, written by someone else. If you now have additionally have to assume that a simple plain or dot access to an object can have effects out of scope it can turn into a pain. I already have the experiences when contributor uses `T&`.  

Finally all that is weighing up: 
* Pain and dot access dos not effect calling code 
vs:
* Easy rule to have right amount of null pointer checks
